src/System.Collections/src/System/Collections/Generic/BitHelper.cs

```diff
@@ -4,136 +4,38 @@
-    /// IsMarked -- pass in values as indices into the logical bit array, and it will be mapped
-    /// to the position within the array of ints.
-    /// 
-    /// FUTURE OPTIMIZATIONS:
-    /// A method such as FindFirstMarked/Unmarked Bit would be useful for callers that operate 
-    /// on a bit array and then need to loop over it. In particular, if it avoided visiting 
-    /// every bit, it would allow good perf improvements when the bit array is sparse.
-    /// </summary>
-    internal unsafe sealed class BitHelper
-    {   // should not be serialized
-        private const byte MarkedBitFlag = 1;
-        private const byte IntSize = 32;
+    internal ref struct BitHelper
+    {
+        private const int IntSize = sizeof(int) * 8;
+        private readonly Span<int> _span;
 
-        // m_length of underlying int array (not logical bit array)
-        private readonly int _length;
-
-        // ptr to stack alloc'd array of ints
-        private readonly int* _arrayPtr;
-
-        // array of ints
-        private readonly int[] _array;
-
-        // whether to operate on stack alloc'd or heap alloc'd array 
-        private readonly bool _useStackAlloc;
-
-        /// <summary>
-        /// Instantiates a BitHelper with a heap alloc'd array of ints
-        /// </summary>
-        /// <param name="bitArrayPtr">Pointer to int array to hold bits</param>
-        /// <param name="length">length of int array</param>
-        internal BitHelper(int* bitArrayPtr, int length)
-        {
-            _arrayPtr = bitArrayPtr;
-            _length = length;
-            _useStackAlloc = true;
-        }
-
-        /// <summary>
-        /// Instantiates a BitHelper with a heap alloc'd array of ints
-        /// </summary>
-        /// <param name="bitArray">int array to hold bits</param>
-        /// <param name="length">length of int array</param>
-        internal BitHelper(int[] bitArray, int length)
+        internal BitHelper(Span<int> span, bool clear)
         {
-            _array = bitArray;
-            _length = length;
+            if (clear)
+            {
+                span.Clear();
+            }
+            _span = span;
         }
 
-        /// <summary>
-        /// Mark bit at specified position
-        /// </summary>
-        /// <param name="bitPosition"></param>
         internal void MarkBit(int bitPosition)
         {
             int bitArrayIndex = bitPosition / IntSize;
-            if (bitArrayIndex < _length && bitArrayIndex >= 0)
+            if ((uint)bitArrayIndex < (uint)_span.Length)
             {
-                int flag = (MarkedBitFlag << (bitPosition % IntSize));
-                if (_useStackAlloc)
-                {
-                    _arrayPtr[bitArrayIndex] |= flag;
-                }
-                else
-                {
-                    _array[bitArrayIndex] |= flag;
-                }
+                _span[bitArrayIndex] |= (1 << (bitPosition % IntSize));
```

Comment:
Since there's no loop, would it be better to do one Math.DivRem instead of separate div and mod?

Reply:
The denominator in this case is const power of two; the JIT just uses shifts to implement the div and mod.